services/worker-manager/src/providers/provider.js

```diff
@@ -84,6 +84,46 @@ class Provider {
     throw new ApiError('not supported for this provider');
   }
 
+  /**
+   * Spawned workers are expected to:
+   * 1. Start (instance is running)
+   * 2. Register (worker is registered with worker manager)
+   * 3. Do work (call queue.claimWork/queue.reclaimTask)
+   *
+   * If worker does not register within given timeout, it will be removed after `terminateAfter` time.
+   * If worker fails to call queue.claimWork, `queue_worker.first_claim` would be set to null.
+   * If worker does not call queue.reclaimTask or stops calling queue.claimWork,
+   * `queue_worker.last_date_active` would not be updated.
+   *
+   * Workers that are registered, but don't have `first_claim`
+   * or `last_date_active` is older than claimTimeout are considered to be zombies.
+   */
+  static isZombie({ worker }) {
+    const claimTimeout = worker.providerData?.claimTimeout || 7200; // last active within 2 hours
+
+    const lastActiveAfter = Date.now() - claimTimeout * 1000;
+
+    let reason = null;
+    let isZombie = false;
+
+    if (!worker.firstClaim && worker.created < lastActiveAfter) {
+      isZombie = true;
+      reason = `worker never claimed work, created=${worker.created}`;
+    }
+
+    if (!worker.lastDateActive && worker.firstClaim < lastActiveAfter) {
```

Comment:
what sets lastDateActive? i'm keen to know what active means in this context.

Reply:
this is something queue service keeps track of for each worker. Whenever worker calls queue methods, those values are being updated. 

Same way we know that worker never called queue service - those values would be missing.